mmaction/models/losses/ssn_loss.py

```diff
@@ -0,0 +1,183 @@
+        """
+        pred = bbox_pred[regression_indexer, :, :]
+        gt = labels[regression_indexer]
+        reg_target = bbox_targets[regression_indexer, :]
+
+        class_idx = gt.data - 1
+        classwise_pred = pred[:, class_idx, :]
+        classwise_reg_pred = torch.cat(
+            (torch.diag(classwise_pred[:, :, 0]).view(
+                -1, 1), torch.diag(classwise_pred[:, :, 1]).view(-1, 1)),
+            dim=1)
+        loss = F.smooth_l1_loss(
+            classwise_reg_pred.view(-1), reg_target.view(-1)) * 2
+        return loss
+
+    def forward(self, activity_score, completeness_score, bbox_pred,
+                proposal_type, labels, bbox_targets, train_cfg):
+        """Calculate Boundary Matching Network Loss.
+
+        Args:
+            activity_score (torch.Tensor): Predicted activity score.
+            completeness_score (torch.Tensor): Predicted completeness score.
+            bbox_pred (torch.Tensor): Predicted interval center and span
+                of positive proposals.
+            proposal_type (torch.Tensor): Type index slices of proposals.
+            labels (torch.Tensor): Groundtruth class label.
+            bbox_targets (torch.Tensor): Groundtruth center and span
+                of positive proposals.
+            train_cfg (dict): Config for training.
+
+        Returns:
+            dict([torch.Tensor, torch.Tensor, torch.Tensor]):
+                (loss_activity, loss_completeness, loss_reg).
+                Loss_activity is the activity loss, loss_completeness is
+                the class-wise completeness loss,
+                loss_reg is the class-wise regression loss.
+        """
+        self.sampler = train_cfg.ssn.sampler
+        self.loss_weight = train_cfg.ssn.loss_weight
+        losses = dict()
+
+        proposal_type = proposal_type.view(-1)
+        labels = labels.view(-1)
+        activity_indexer = ((proposal_type == 0) +
+                            (proposal_type == 2)).nonzero().squeeze(1)
+        completeness_indexer = ((proposal_type == 0) +
+                                (proposal_type == 1)).nonzero().squeeze(1)
+
+        total_ratio = (
+            self.sampler.positive_ratio + self.sampler.background_ratio +
+            self.sampler.incomplete_ratio)
+        positive_per_video = int(self.sampler.num_per_video *
+                                 (self.sampler.positive_ratio / total_ratio))
+        background_per_video = int(
+            self.sampler.num_per_video *
+            (self.sampler.background_ratio / total_ratio))
+        incomplete_per_video = (
+            self.sampler.num_per_video - positive_per_video -
+            background_per_video)
+
+        losses['loss_activity'] = self.activity_loss(activity_score, labels,
+                                                     activity_indexer)
+        losses['loss_activity'] = torch.tensor(
+            losses['loss_activity'], device=activity_score.device)
```

Comment:
This line is redundant since the return of `self.activity_loss()` already in activity_score device.